File: samples/SampleTodo.Service/Events.cs
using Metano.Annotations;
using Metano.Annotations.TypeScript;

namespace SampleTodo.Service;

/// <summary>
/// Canonical kinds of events the todo service surfaces. <c>[StringEnum]</c>
/// lowers this to a TS string union so the discriminator narrowing
/// compiles to a literal comparison.
/// </summary>
[Transpile, StringEnum, EmitInFile("events")]
public enum TodoEventKind
{
    TodoCreated,
    TodoUpdated,
    TodoDeleted,
}

/// <summary>
/// Event published when a new todo is inserted. <c>[Discriminator]</c>
/// marks <c>Kind</c> as the narrowing field — the generated
/// <c>isTodoCreated</c> guard short-circuits on <c>v.kind !== "TodoCreated"</c>
/// before walking the rest of the shape, so a payload for a different
/// event exits the guard without scanning every field.
/// </summary>
[PlainObject, GenerateGuard, Discriminator("Kind"), EmitInFile("events")]
public record TodoCreated(TodoEventKind Kind, string Id, string Title);

/// <summary>
/// Event published when a todo is patched. Shares the same
/// <see cref="TodoEventKind"/> discriminator, different expected value
/// (<c>"TodoUpdated"</c>) — the emitted guard keyword matches the type
/// name by convention.
/// </summary>
[PlainObject, GenerateGuard, Discriminator("Kind"), EmitInFile("events")]
public record TodoUpdated(TodoEventKind Kind, string Id, string? Title, bool? Completed);

File: targets/js/sample-todo-service/src/events.ts
export const TodoEventKind = {
  TodoCreated: "TodoCreated",
  TodoUpdated: "TodoUpdated",
  TodoDeleted: "TodoDeleted",
} as const;

export type TodoEventKind = typeof TodoEventKind[keyof typeof TodoEventKind];

export interface TodoCreated {
  readonly kind: TodoEventKind;
  readonly id: string;
  readonly title: string;
}

export function isTodoCreated(value: unknown): value is TodoCreated {
  if (value == null || typeof value !== "object") {
    return false;
  }

  const v = value as any;

  if (v.kind !== "TodoCreated") {
    return false;
  }

  return typeof v.id === "string" && typeof v.title === "string";
}

export function assertTodoCreated(value: unknown, message?: string): asserts value is TodoCreated {
  if (!isTodoCreated(value)) {
    throw new TypeError(message ?? "Value is not a TodoCreated");
  }
}

export interface TodoUpdated {
  readonly kind: TodoEventKind;
  readonly id: string;
  readonly title: string | null;
  readonly completed: boolean | null;
}

export function isTodoUpdated(value: unknown): value is TodoUpdated {
  if (value == null || typeof value !== "object") {
    return false;
  }

  const v = value as any;

  if (v.kind !== "TodoUpdated") {
    return false;
  }

  return typeof v.id === "string" && (v.title == null || typeof v.title === "string") && (v.completed == null || typeof v.completed === "boolean");
}

export function assertTodoUpdated(value: unknown, message?: string): asserts value is TodoUpdated {
  if (!isTodoUpdated(value)) {
    throw new TypeError(message ?? "Value is not a TodoUpdated");
  }
}
